src/Std/Data/DTreeMap/Internal/Zipper.lean

```diff
@@ -0,0 +1,1152 @@
+    | .eq => .cons k v r it
+    | .gt => prependMapGE r lower_bound it
+
+public def Zipper.prependMapGT [Ord α] (t : Impl α β) (lower_bound : α)
+    (it : Zipper α β) : Zipper α β :=
+  match t with
+  | .leaf => it
+  | .inner _ k v l r =>
+    match compare lower_bound k with
+    | .lt => prependMapGT l lower_bound (Zipper.cons k v r it)
+    | .eq => prependMapGT r lower_bound it
+    | .gt => prependMapGT r lower_bound it
+
+theorem Zipper.prepend_prune_LE_eq_prependMapGE [Ord α] (t : Impl α β) (ord_t : t.Ordered)
+    (z : Zipper α β) (lower_bound : α) :
+    z.prependMap (t.prune_LE ord_t lower_bound) = z.prependMapGE t lower_bound := by
+  induction t generalizing z
+  case leaf =>
+    simp only [Impl.prune_LE, Zipper.prependMap, Zipper.prependMapGE]
+  case inner _ k v l r l_ih r_ih =>
+    simp only [Impl.prune_LE, Zipper.prependMapGE]
+    generalize heq : compare lower_bound k = x
+    cases x
+    case lt =>
+      simp only
+      apply l_ih
+    case eq =>
+      simp only [Zipper.prependMap]
+    case gt =>
+      simp only
+      apply r_ih
+
+theorem Zipper.prepend_eq_prependMapGT_self [Ord α] [TransOrd α] (r : Impl α β)
+    (z : Zipper α β) (lower_bound : α) (ord_r : r.Ordered)
+    (hyp : ∀ e ∈ r.toList, compare lower_bound e.fst = .lt) :
+    Zipper.prependMap r z = Zipper.prependMapGT r lower_bound z := by
+  induction r generalizing z
+  case leaf =>
+    trivial
+  case inner _ k v l r l_ih r_ih =>
+    simp only [Zipper.prependMap, Zipper.prependMapGT]
+    have hyp' := hyp ⟨k,v⟩ (by simp only [Impl.toList_eq_toListModel,
+      Impl.toListModel_inner, List.mem_append, List.mem_cons, true_or, or_true])
+    simp at hyp'
+    rw [hyp']
+    apply l_ih
+    . exact (Impl.Ordered.left ord_r)
+    . intro e mem
+      apply hyp e
+      simp only [Impl.toList_eq_toListModel, Impl.toListModel_inner,
+        List.mem_append, List.mem_cons]
+      apply Or.inl
+      . rw [Impl.toList_eq_toListModel] at mem
+        exact mem
+
+theorem Zipper.prepend_prune_LT_eq_prependMapGT [Ord α] [TransOrd α] (t : Impl α β)
+    (ord_t : t.Ordered) (z : Zipper α β) (lower_bound : α) :
+    z.prependMap (t.prune_LT ord_t lower_bound) = z.prependMapGT t lower_bound := by
+  induction t generalizing z
+  case leaf =>
+    simp only [Impl.prune_LT, Zipper.prependMap, Zipper.prependMapGT]
+  case inner _ k v l r l_ih r_ih =>
+    simp only [Impl.prune_LT, Zipper.prependMapGT]
+    generalize heq : compare lower_bound k = x
+    cases x
+    case lt =>
+      simp only [Zipper.prependMap]
+      apply l_ih
+    case eq =>
+      simp only
+      apply prepend_eq_prependMapGT_self
+      . exact Impl.Ordered.right ord_t
+      . intro e mem
+        apply TransCmp.lt_of_eq_of_lt heq
+        apply Impl.Ordered.compare_right ord_t
+        rw [← Impl.toList_eq_toListModel]
+        exact mem
+    case gt =>
+      simp
+      apply r_ih
+
+public theorem Zipper.prependMap_toList_eq_concat_toList (t : Impl α β)
```

Comment:
As per the naming convention, this should be called `toList_prependMap_eq_append` or just `toList_prependMap` would also be fine.